akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala

```diff
@@ -1709,6 +1709,11 @@ final class Flow[In, Out, Mat](delegate: scaladsl.Flow[In, Out, Mat]) extends Gr
    * a new substream is opened and subsequently fed with all elements belonging to
    * that key.
    *
+   * WARNING: If `allowClosedSubstreamRecreation` is set to `false` (default behavior) the stage
+   * keeps track of all keys of streams that have already been closed. If you expect an infinite
+   * number of keys this can cause memory issues. Elements belonging to those keys are drained
+   * directly and not send to the substream.
```

Comment:
Shouldn't we also warn about the semantics of `allowClosedSubstreamRecreation=true` as discussed in the other PR:

https://github.com/akka/akka/pull/24783#issuecomment-376769171

https://github.com/akka/akka/pull/24783#issuecomment-376794466

Reply:
Hmm, at least a comment on the racyness would be appropriate I think. I had to circumvent that problem in the tests as well.

```scala
Source(1, 2, 1, 1).groupBy(Int.MaxValue, identity, true).take(1).mergeSubstreams.runForeach(println)
```

Isn't guaranteed to print all the elements for instance.